src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb

```diff
@@ -0,0 +1,993 @@
+    Partial Class BoundCaseClause
+        Implements ICaseClause
+
+        Protected MustOverride ReadOnly Property ICaseClass As CaseKind Implements ICaseClause.CaseKind
+    End Class
+
+    Partial Class BoundSimpleCaseClause
+        Implements ISingleValueCaseClause
+
+        Private ReadOnly Property IEquality As BinaryOperationKind Implements ISingleValueCaseClause.Equality
+            Get
+                ' Can lifted operators appear here, and if so what is their correct treatment?
+                Dim caseValue As BoundExpression = DirectCast(Me.IValue, BoundExpression)
+                If caseValue IsNot Nothing Then
+                    Select Case caseValue.Type.SpecialType
+                        Case SpecialType.System_Int32, SpecialType.System_Int64, SpecialType.System_UInt32, SpecialType.System_UInt64, SpecialType.System_UInt16, SpecialType.System_Int16, SpecialType.System_SByte, SpecialType.System_Byte, SpecialType.System_Char
+                            Return BinaryOperationKind.IntegerEquals
+
+                        Case SpecialType.System_Boolean
+                            Return BinaryOperationKind.BooleanEquals
+
+                        Case SpecialType.System_String
+                            Return BinaryOperationKind.StringEquals
+                    End Select
+
+                    If caseValue.Type.TypeKind = TypeKind.Enum Then
+                        Return BinaryOperationKind.EnumEquals
+                    End If
+                End If
+
+                Return BinaryOperationKind.None
+            End Get
+        End Property
+
+        Private ReadOnly Property IValue As IExpression Implements ISingleValueCaseClause.Value
+            Get
+                If Me.ValueOpt IsNot Nothing Then
+                    Return Me.ValueOpt
+                End If
+
+                If Me.ConditionOpt IsNot Nothing AndAlso Me.ConditionOpt.Kind = BoundKind.BinaryOperator Then
+                    Dim value As BoundBinaryOperator = DirectCast(Me.ConditionOpt, BoundBinaryOperator)
+                    If value.OperatorKind = BinaryOperatorKind.Equals Then
+                        Return value.Right
+                    End If
+                End If
+
+                Return Nothing
+            End Get
+        End Property
+
+        Protected Overrides ReadOnly Property ICaseClass As CaseKind
+            Get
+                Return If(Me.IValue IsNot Nothing, CaseKind.SingleValue, CaseKind.Default)
+            End Get
+        End Property
+    End Class
+
+    Partial Class BoundRangeCaseClause
+        Implements IRangeCaseClause
+
+        Private ReadOnly Property IMaximumValue As IExpression Implements IRangeCaseClause.MaximumValue
+            Get
+                If Me.UpperBoundOpt IsNot Nothing Then
+                    Return Me.UpperBoundOpt
+                End If
+
+                If Me.UpperBoundConditionOpt IsNot Nothing AndAlso Me.UpperBoundConditionOpt.Kind = BoundKind.BinaryOperator Then
```

Comment:
If UpperBoundOpt is null, UpperBoundConditionOpt is non-null. Same for LowerBoundConditionOpt below.